Add auction item

rstein66 opened the pull request:
- Item has a title
  * title cannot be blank
- Item has a description 
  * description cannot be blank
- Item belongs to a user
- User sets a price
  * price should be greater than 0
  * price should be displayed in $ (dollars)
  * price should support cents
- User sets an end date for the auction
  * end date should be in the future
- The user should be able to correct invalid data

xray committed:
add money dependency

xray committed:
add migration and schema for auction

xray committed:
add interface to create auction

rstein66 committed:
Add editorconfig; extend gitignore

rstein66 committed:
Minor cleanup- fix typos, remove unnecessary, etc

xray committed:
remove unnecessary label

xray committed:
remove unused alias

rstein66 committed:
Add user-auction association

rstein66 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[auctions] User can create an auction for an item. (6.25pt)](https://trello.com/c/q4kMdA03/4-auctions-user-can-create-an-auction-for-an-item-625pt)

kevbuchanan commented on lib/renaissance_web/router.ex:
```diff
@@ -21,6 +21,8 @@ defmodule RenaissanceWeb.Router do
     get "/login", LoginController, :login
     post "/login", LoginController, :verify
+    get "/auctions/new", AuctionController, :new
+    post "/auctions/new", AuctionController, :create
```
This is typically `post "/auctions"`:

https://hexdocs.pm/phoenix/routing.html#resources

rstein66 replied:
Resolved with https://github.com/xray/Renaissance/commit/9a52acc90840caae1ee2068273d642fd86942a52

kevbuchanan commented on .gitignore:
```diff
@@ -175,3 +175,73 @@ modules.xml
 /installer/deps
 
 # End of https://www.gitignore.io/api/phoenix
+
```
This is getting a little out of hand. It's unreasonable to think that every kind of file might be captured in the `.gitignore`. Best to have a more disciplined git workflow and review process.

ssunday commented on .editorconfig:
```diff
@@ -0,0 +1,16 @@
+root = true
+
+[*]
+indent_style = space
```
What is this for..?

ssunday replied:
I mean, should this be checked in period

kevbuchanan commented on lib/renaissance/auctions/auction.ex:
```diff
@@ -0,0 +1,48 @@
+    field :description, :string
+    field :price, Money.Ecto.Amount.Type
+    field :end_date, :utc_datetime
+    belongs_to :user, User
```
Getting this warning here:

> warning: invalid association `user` in schema Renaissance.Auction: associated schema User does not exist
  lib/renaissance/auctions/auction.ex:1: Renaissance.Auction (module)

ssunday commented on assets/css/app.css:
```diff
@@ -66,10 +66,37 @@ body {
+}
+
+.half-right {
+    margin: 0 0 0 1rem;
```
isn't this just margin-left: 1rem;?

rstein66 replied:
Resolved with 92244839629182a73fc9096c4664cc861b65b1a2

ssunday commented on lib/renaissance/auctions/auctions.ex:
```diff
@@ -0,0 +1,48 @@
+
+  defp format_end_date(params) do
+    datetime =
+      "#{params["end_date_day"]}T#{params["end_date_time"]}:00-05:00"
```
is this just formatting it in CST because this looks very fragile

ssunday replied:
Er, CDT, since -5, I think

rstein66 replied:
Resolved with 159a775e493da7136d042a4f76b690b4d8cb2aea

ssunday commented on lib/renaissance_web/router.ex:
```diff
@@ -21,6 +21,8 @@ defmodule RenaissanceWeb.Router do
     post "/register", RegisterController, :register
     get "/login", LoginController, :login
     post "/login", LoginController, :verify
+    get "/auctions/new", AuctionController, :new
```
Usually posting to index is the create route, thus you could use resources

ssunday commented on test/renaissance/auctions/auction_test.exs:
```diff
@@ -0,0 +1,139 @@
+      changeset =
+        Auction.changeset(
+          %Auction{},
+          %{
```
Could you extract out the base-line attrs into global so you can re-use and then change it as necessary?

rstein66 replied:
Resolved with 159a775e493da7136d042a4f76b690b4d8cb2aea

kevbuchanan commented on lib/renaissance/users/user.ex:
```diff
@@ -2,33 +2,33 @@ defmodule Renaissance.User do
-      _ ->
-        hash = Bcrypt.hash_pwd_salt(password)
-        attrs |> put_change(:password_hash, hash)
+    if is_nil(password = get_change(attrs, :password)) do
```
The assignment in here is a little unconventional and unclear. I'm kind of surprised it's not a compile warning: https://github.com/elixir-lang/elixir/issues/7403

rstein66 replied:
I understood this to mean that it's unconventional & unclear because assignment occurs in the conditional, and so I addressed that here https://github.com/xray/Renaissance/commit/159a775e493da7136d042a4f76b690b4d8cb2aea#diff-3e2a5dd6d0ebb3b2d8b31401bf178663L28.

**If I misunderstood and did not address this, please let me know.**

kevbuchanan commented on lib/renaissance_web/controllers/auction_controller.ex:
```diff
@@ -0,0 +1,33 @@
+  def create(conn, params) do
+    changeset =
+      conn
+      |> get_session(:current_user)
```
I guess it wasn't clear to me in the original PR, but I would not have expected `session[:current_user]` to be the email address. I would have expected something like `get_session(:current_user_id)`. Just use the primary key, then you also avoid having to look up id from email wherever you want to attach something to the current user.

ssunday replied:
agree

kevbuchanan commented on priv/repo/migrations/20190413180607_create_auctions.exs:
```diff
@@ -0,0 +1,15 @@
+
+  def change do
+    create table(:auctions) do
+      add :title, :string
```
Probably best to also add `null: false` to the database schema if these are required.

ssunday replied:
I'd only do that for the super core fields, not the aux ones

kevbuchanan commented on test/renaissance/auctions/auction_test.exs:
```diff
@@ -0,0 +1,139 @@
+      assert %{description: ["can't be blank"]} = errors_on(changeset)
+    end
+
+    test "requires user id" do
```
It would be nice to have a test for an invalid user_id (one that isn't a valid foreign key based on the `references(:users)` foreign key constraint.

kevbuchanan commented on lib/renaissance/auctions/auctions.ex:
```diff
@@ -0,0 +1,48 @@
+    datetime =
+      "#{params["end_date_day"]}T#{params["end_date_time"]}:00-05:00"
+      |> DateTime.from_iso8601()
+      |> elem(1)
```
I think you're using `elem(1)` here to avoid handling the result tuple of either:
`{:ok, date}`, `{:error, reason}`. In which case if you actually want to not handle the error case you can use `DateTime.from_iso8601!()`

rstein66 replied:
Resolved with 159a775e493da7136d042a4f76b690b4d8cb2aea

rstein66 committed:
Remove editorconfig; remove gitignore extension

Partial reversion of commit
https://github.com/xray/Renaissance/commit/0e833eaac725dea32a1a5475036a7e336b2a81b3
 motivation
- https://github.com/xray/Renaissance/pull/3#discussion_r275946850
- https://github.com/xray/Renaissance/pull/3#discussion_r275948021

rstein66 committed:
Ensure core fields [in migration] are not null

ssunday commented on assets/css/app.css:
```diff
@@ -92,6 +119,23 @@ body {
+    width: 50%;
+}
+
+.dollar-sign{
```
missing space

rstein66 replied:
Resolved with 92244839629182a73fc9096c4664cc861b65b1a2

ssunday commented on lib/renaissance/auctions/auction.ex:
```diff
@@ -0,0 +1,48 @@
+    zero_dollars = Money.new(000, :USD)
+    one_dollar = Money.new(100, :USD)
+
+    if Money.compare(item_price || one_dollar, zero_dollars) == 1 do
```
Looking at this, it isn't clear what it means IMO

ssunday replied:
I mean the comparison statement, could you extract it out?

rstein66 replied:
Resolved with 159a775e493da7136d042a4f76b690b4d8cb2aea

rstein66 commented:
Duplicated by https://github.com/xray/Renaissance/pull/4 => 
Feedback [is/will be] incorporated into https://github.com/xray/Renaissance/pull/4
